enigma-tools-m/src/primitives/km_primitives.rs

```diff
@@ -0,0 +1,263 @@
+use enigma_types::{ContractAddress, DhKey, PubKey, StateKey};
+
+pub type MsgID = [u8; 12];
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum PrincipalMessageType {
+    Response(Vec<(ContractAddress, StateKey)>),
+    Request(Option<Vec<ContractAddress>>),
+    EncryptedResponse(Vec<u8>),
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct PrincipalMessage {
+    pub data: PrincipalMessageType,
+    pub(crate) pubkey: Vec<u8>,
+    pub(crate) id: MsgID,
+}
+
+impl PrincipalMessage {
+    pub fn new(data: PrincipalMessageType, pubkey: PubKey) -> Result<Self, CryptoError> {
+        let mut id = [0u8; 12];
+        rand::random(&mut id)?;
+        let pubkey = pubkey.to_vec();
+        Ok(Self { data, pubkey, id })
+    }
+
+    pub fn new_id(data: PrincipalMessageType, id: [u8; 12], pubkey: PubKey) -> Self {
+        let pubkey = pubkey.to_vec();
+        Self { data, pubkey, id }
+    }
+
+    pub fn to_sign(&self) -> Result<Vec<u8>, ToolsError> {
+        if self.is_response() {
+            return Err(MessagingError { err: "can't serialize non encrypted response" });
+        }
+        let mut to_sign = Vec::with_capacity(3);
+        match &self.data {
+            PrincipalMessageType::Request(Some(addresses)) => {
+                to_sign.push(hash::prepare_hash_multiple(addresses.as_ref()));
+            }
+            PrincipalMessageType::EncryptedResponse(v) => to_sign.push(v.clone()),
+            PrincipalMessageType::Request(None) => (), // If the request is empty we don't need to sign on it.
+            PrincipalMessageType::Response(_) => unreachable!(), // This can't be reached because we check if it's a response before.
+        }
+        to_sign.push(self.pubkey.to_vec());
+        to_sign.push(self.id.to_vec());
+        Ok(hash::prepare_hash_multiple(&to_sign))
+    }
+
+    pub fn into_message(self) -> Result<Vec<u8>, ToolsError> {
+        if self.is_response() {
+            return Err(MessagingError { err: "can't serialize non encrypted response" });
+        }
+        let mut buf = Vec::new();
+        let val = serde_json::to_value(self).map_err(|_| MessagingError { err: "can't serialize the message" })?;
+        val.serialize(&mut Serializer::new(&mut buf)).map_err(|_| MessagingError { err: "can't serialize the message" })?;
+        Ok(buf)
+    }
+
+    pub fn from_message(msg: &[u8]) -> Result<Self, ToolsError> {
+        let mut des = Deserializer::new(msg);
+        let res: serde_json::Value =
+            Deserialize::deserialize(&mut des).map_err(|_| MessagingError { err: "can't deserialize the message" })?;
+        let msg: Self = serde_json::from_value(res).map_err(|_| MessagingError { err: "can't deserialize the message" })?;
+        Ok(msg)
+    }
+
+    pub fn get_pubkey(&self) -> PubKey {
+        let mut pubkey = [0u8; 64];
+        pubkey.copy_from_slice(&self.pubkey[..]);
+        pubkey
+    }
+
+    pub fn get_id(&self) -> MsgID { self.id }
+
+    pub fn is_request(&self) -> bool {
```

Comment:
I didn't see that this `is_<type>` functions are used anywhere. If they are indeed not used, it is better to remove them.

Reply:
`is_response` is used ( https://github.com/enigmampc/enigma-core/blob/159a9b6c2f3e444b9380ad5da13ed4173d772a51/enigma-tools-m/src/primitives/km_primitives.rs#L39 ) but `is_request`/`is_encrypted_response` aren't used.